### DISCUSSION TITLE
How do I add config fields for XrayJsonReporter dynamically?

### DISCUSSION THREAD

**Discussion by kamil-nowocin** [Xray-App/xray-testng-extensions#15] [General]
I would like these fields to be created dynamically. I have seen a solution with the `xray.properties` file

```
user=xpto
project_key=CALC
summary=test automation results
description=selenium test results
version=1.0
revision=123
testplan_key=CALC-1000
test_environments=chrome
report_filename=xray-report.json
```

but I would like to create them dynamically and pass them on. How can I implement this? Assuming I have the following configuration file 
```
public class XrayJsonConfig {

  public static XrayJsonReporterConfig getXrayJsonReporterConfig() {
    var config = new XrayJsonReporterConfig();
    config.setSummary("Test description %s".formatted(EnvironmentManager.getAppReleaseId()));
    config.setTestPlanKey("XXXXX-18513");
    config.setReportFilename("xray-report.json");
    config.setProjectKey("XXXXX");
    config.setDescription("""
        Collection of automated tests performed during testing of the application with version %s.
        Allure Report: %s
        CI/CD Execution: %s
        """.formatted(EnvironmentManager.getAppReleaseId(), "https://test.com",
        "https://test.com"));
    config.setUser("Jira User");
    return config;
  }
}
```

How do I pass this on to XrayJsonReporter.class to make it work?
I also added this class as listener
```
@Listeners({TestListener.class, AnnotationTransformer.class, MethodListener.class,
    XrayListener.class, XrayJsonReporter.class})
public class BaseHooks
```

and test class
```
public class ExperimentalTest extends BaseHooks {

  @XrayTest(key = "XXXXX-123", labels = "LABEL")
  @Test(testName = "EXPERIMENTAL TEST - NAME", description = "EXPERIMENTAL TEST - DESC")
  public void experimentalTest() {
    // ACT
    log.info("TEST LOG!");
  }
}
````

**Comment by bitcoder** [comment]
I haven't implemented an explicit way to handle it, as far as I remember it.
I'm not sure if there's a specific way of doing it officialy with TestNG; some investigation needs to be performed.
Suggestions/feedback are welcome.

**Comment by kamil-nowocin** [comment]
Unfortunately, I have no idea how to get it right. However, it would be quite an improvement/interesting functionality.

**Comment by Atari361** [comment]
I also had similar need when implementing Xray support for my project, and solved this by introducing an extension class to `XrayJsonReporter` and use that class as listener instead of `XrayJsonReporter`. This way you now have access to public getConfig() method from TestNG's IReporter interface and finally can set report params in runtime, here is a simplified example:
```
public class XrayJsonReporterExtension extends XrayJsonReporter {

    @Override
    public void generateReport(List<XmlSuite> xmlSuites, List<ISuite> suites, String outputDirectory) {
        setXraySettings();
        super.generateReport(xmlSuites, suites, outputDirectory);
    }

    private void setXraySettings() {
        getConfig().setXrayCloud(true);
        getConfig().setProjectKey("DEMO");
        getConfig().setUser(ProjectProps.getProperty(JIRA_USER_EMAIL));
        getConfig().setSummary("Regression results");
        getConfig().setTestPlanKey(ProjectProps.getProperty(XRAY_TEST_PLAN));
        getConfig().setDescription("*Backend*: %s\n*Frontend*: %s"
                .formatted(getBackendVersion(), getFrontEndVersion()));
    }
}
```
config itself in XrayJsonReporter is final so there seems to be no way to set it it the way you thought it would be (your method `getXrayJsonReporterConfig()`)

**Reply by kamil-nowocin** [reply]
In the end, I gave up using this plugin, implemented my own listener to generate a report in JSON format accepted by the Jira Xray plugin

**Reply by bitcoder** [reply]
> I also had similar need when implementing Xray support for my project, and solved this by introducing an extension class to `XrayJsonReporter` and use that class as listener instead of `XrayJsonReporter`. This way you now have access to public getConfig() method from TestNG's IReporter interface and finally can set report params in runtime, here is a simplified example:
> 
> ```
> public class XrayJsonReporterExtension extends XrayJsonReporter {
> 
>     @Override
>     public void generateReport(List<XmlSuite> xmlSuites, List<ISuite> suites, String outputDirectory) {
>         setXraySettings();
>         super.generateReport(xmlSuites, suites, outputDirectory);
>     }
> 
>     private void setXraySettings() {
>         getConfig().setXrayCloud(true);
>         getConfig().setProjectKey("DEMO");
>         getConfig().setUser(ProjectProps.getProperty(JIRA_USER_EMAIL));
>         getConfig().setSummary("Regression results");
>         getConfig().setTestPlanKey(ProjectProps.getProperty(XRAY_TEST_PLAN));
>         getConfig().setDescription("*Backend*: %s\n*Frontend*: %s"
>                 .formatted(getBackendVersion(), getFrontEndVersion()));
>     }
> }
> ```
> 
> config itself in XrayJsonReporter is final so there seems to be no way to set it it the way you thought it would be (your method `getXrayJsonReporterConfig()`)

a PR would be appreciated to enhance the current extension

**Comment by bitcoder** [comment]
After some analysis, here's my feedback:

To be able to set configuration parameters at runtime, I see 2 options:
1) extend the `XrayJsonReporter` class, similar to what @Atari361 mentioned, and use `getConfig()` to set properties; use the new class as the reporter
2) execute the tests programmaticaly, creating at `TestNG` object, creating and configuring the listeners, and running the tests

```java
        TestNG testng = new TestNG();
        List<XmlClass> classes = new ArrayList<XmlClass>();
        XmlSuite suite = new XmlSuite();
        suite.setName("XrayTests");
        XmlTest test = new XmlTest(suite);
        test.setName("ReportTest");
           
        XmlClass class1 = new XmlClass(testClass);
        class1.setIncludedMethods(Collections.singletonList(new XmlInclude(methodName)));
         classes.add(class1);
        test.setXmlClasses(classes);  
        testng.setXmlSuites(Collections.singletonList(suite));
 

        List<java.lang.Class<? extends ITestNGListener>> listeners = new ArrayList<>();
        testng.setListenerClasses(listeners);
        XrayJsonReporter reporter = new XrayJsonReporter();

        // set configuration parameters...
        reporter.getConfig(). setTestPlanKey('...');
     

        testng.addListener(reporter);


        testng.run();
```

TestNG [initializes](https://github.com/testng-team/testng/blob/0f779fe8a565d5e1024f34bc49e8a1c3df1ed8a7/testng-core/src/main/java/org/testng/TestNG.java#L1812) the reporter and then invokes the `getConfig()`  method using the `ReporterConfig` object.
I have not seen any other way of changing it.

Anyway, since the idea is to be able to set these parameters easily on the pipeline without having to use predefined values on a .properties file, the current reporter should  be able to get these parameters from environment variables that we can then refer to on the pom.xml file.

As a recap, we can also:
 - configure/hardocde parameters for the reporter in the pom.xml file as mentioned on the README, and surefire will take care of passing them to the reporter using setters; here we can eventually fill the values from environment variables
 - configure/hardcode parameters on a properties file

I'm not sure if we can get access to the running listeners at runtime and then eventually invoke the getConfig() to set parameters... that would be perhaps a bit cumbersome.. but maybe it could work as the generation of the report using the `generateReport()`  method is only called at the "end".
